### PR TITLE
accounts, cmd, internal: disable unlock account as default

### PR DESCRIPTION
**Pull request by rjl493456442**
Implements #17023

### CODE REVIEW

File: cmd/geth/main.go [hunk 1]
```diff
@@ -325,3 +319,32 @@ func startNode(ctx *cli.Context, stack *node.Node) {
...
+		ks := stack.AccountManager().Backends(keystore.KeyStoreType)[0].(*keystore.KeyStore)
+		passwords := utils.MakePasswordList(ctx)
+		unlocks := strings.Split(ctx.GlobalString(utils.UnlockedAccountFlag.Name), ",")
+		for i, account := range unlocks {
+			if trimmed := strings.TrimSpace(account); trimmed != "" {
+				unlockAccount(ctx, ks, trimmed, i, passwords)
+			}
+		}
+	} else {
+		log.Info("Not safe to unlock accounts. Please explicitly enable it if still need to unlock")
```
Comment (ligi) [inline cmd/geth/main.go]: perhaps mention the flag to do so here for better UX

File: cmd/geth/main.go [hunk 2]
```diff
@@ -325,3 +319,32 @@ func startNode(ctx *cli.Context, stack *node.Node) {
...
+			if module == "personal" {
+				return true
+			}
+		}
+		return false
+	}
+
+	// If user explicitly enable insecure account unlocking or account-related APIs will
+	// not be exposed by http, execute unlock operation.
+	enable := ctx.GlobalBool(utils.InsecureUnlockFlag.Name)
```
Comment (ligi) [inline cmd/geth/main.go]: nit: I would rename enable to something like isInsecureFlagEnabled

then invert the condition for better readability.

if  !isInsecureFlagEnabled && personalExposed() {
 log
 return
}

File: cmd/geth/main.go (hunk 1 above)
Comment (ligi) [reply cmd/geth/main.go]: and perhaps Info->Warn

File: internal/ethapi/api.go [hunk 3]
```diff
@@ -317,7 +319,14 @@ func (s *PrivateAccountAPI) ImportRawKey(privkey string, password string) (commo
 // UnlockAccount will unlock the account associated with the given address with
 // the given password for duration seconds. If duration is nil it will use a
 // default of 300 seconds. It returns an indication if the account was unlocked.
-func (s *PrivateAccountAPI) UnlockAccount(addr common.Address, password string, duration *uint64) (bool, error) {
+func (s *PrivateAccountAPI) UnlockAccount(ctx context.Context, addr common.Address, password string, duration *uint64) (bool, error) {
+	// When the API is exposed by HTTP, unless the user explicitly specifies to allow the insecure account unlocking,
+	// otherwise it is disabled.
+	isHttp, ok := ctx.Value("http").(bool)
+	if ok && isHttp && !s.unlockAllowed {
+		return false, errors.New("insecure unlock operation is not allowed")
```
Comment (ligi) [inline internal/ethapi/api.go]: also mention the flag here for UX and perhaps even use the same string - DRY

**Comment by rjl493456442** [comment]
@ligi Thanks for your comments, now updated! For the todo, i think we need to decouple the config stuff to fix it. Maybe in another PR will be better.

**Comment by ligi** [comment]
Great - thanks!
looks good - would just really DRY the error message - currently it's in there twice - slightly different

> Not safe to unlock accounts. Please enable `allow-insecure-unlock` flag if still need to unlock

and

> not safe to unlock account, please enable `allow-insecure-unlock` flag if necessary

I would really define this message once somewhere - also has the advantage that it connects both code-parts that care for a similar thing. Also it streamlines things and makes things DRY ;-)

File: cmd/geth/main.go [hunk 4]
```diff
@@ -325,3 +319,31 @@ func startNode(ctx *cli.Context, stack *node.Node) {
...
+	// personalExposed checks whether personal API is exposed by http.
+	personalExposed := func() bool {
+		for _, module := range stack.Config().HTTPModules {
+			if module == "personal" {
+				return true
+			}
+		}
+		return false
+	}
+
```
Comment (ccding) [inline cmd/geth/main.go]: remove empty line

File: cmd/geth/main.go [hunk 5]
```diff
@@ -325,3 +319,31 @@ func startNode(ctx *cli.Context, stack *node.Node) {
...
+	// If insecure account unlocking is not allowed and account-related APIs is exposed by http,
```
Comment (ccding) [inline cmd/geth/main.go]: is -> are

File: cmd/geth/main.go [hunk 6]
```diff
@@ -325,3 +319,31 @@ func startNode(ctx *cli.Context, stack *node.Node) {
 		}
 	}
 }
+
+// unlockAccounts unlocks any account specifically requested.
+func unlockAccounts(ctx *cli.Context, stack *node.Node) {
+	// personalExposed checks whether personal API is exposed by http.
+	personalExposed := func() bool {
```
Comment (ccding) [inline cmd/geth/main.go]: why not just write this as a function?

File: cmd/geth/main.go [hunk 7]
```diff
@@ -325,3 +319,31 @@ func startNode(ctx *cli.Context, stack *node.Node) {
...
+				return true
+			}
+		}
+		return false
+	}
+
+	// If insecure account unlocking is not allowed and account-related APIs is exposed by http,
+	// print warning log to user and skip unlocking.
+	if !stack.Config().InsecureUnlockAllowed && personalExposed() {
+		log.Warn("not safe to unlock account, please enable `allow-insecure-unlock` flag if necessary")
```
Comment (ccding) [inline cmd/geth/main.go]: its better to `return` after the warning, and move the `else` branch outside

File: cmd/geth/main.go (hunk 6 above)
Comment (rjl493456442) [reply cmd/geth/main.go]: I think this function is specialized and may not be invoked by other part codebase. So maybe embedded function is fine here.

**Comment by holiman** [comment]
Sorry, I totally missed this PR. Would you care to fix the conflicts again?

**Comment by rjl493456442** [comment]
@holiman Sure! But this PR is super old, I need some time to recall the implementation detail.

**Comment by rjl493456442** [comment]
@holiman Updated, please take another look.

File: accounts/manager.go [hunk 8]
```diff
@@ -40,7 +45,7 @@ type Manager struct {
 
 // NewManager creates a generic account manager to sign transaction via various
 // supported backends.
-func NewManager(backends ...Backend) *Manager {
+func NewManager(insecureUnlockAllowed bool, backends ...Backend) *Manager {
```
Comment (karalabe) [inline accounts/manager.go]: Perhaps lets pass the config in here, not just the flag.

File: cmd/geth/main.go [hunk 9]
```diff
@@ -398,3 +391,30 @@ func startNode(ctx *cli.Context, stack *node.Node) {
 		}
 	}
 }
+
+// unlockAccounts unlocks any account specifically requested.
+func unlockAccounts(ctx *cli.Context, stack *node.Node) {
+	// personalExposed checks whether personal API is exposed by http.
+	personalExposed := func() bool {
+		for _, module := range stack.Config().HTTPModules {
```
Comment (karalabe) [inline cmd/geth/main.go]: We should not tie it to `personal` rather to the entire RPC (both HTTP and WS). It's safer as `eth` is also susceptible.

File: internal/ethapi/api.go [hunk 10]
```diff
@@ -318,7 +318,13 @@ func (s *PrivateAccountAPI) ImportRawKey(privkey string, password string) (commo
 // UnlockAccount will unlock the account associated with the given address with
 // the given password for duration seconds. If duration is nil it will use a
 // default of 300 seconds. It returns an indication if the account was unlocked.
-func (s *PrivateAccountAPI) UnlockAccount(addr common.Address, password string, duration *uint64) (bool, error) {
+func (s *PrivateAccountAPI) UnlockAccount(ctx context.Context, addr common.Address, password string, duration *uint64) (bool, error) {
+	// When the API is exposed by HTTP, unless the user explicitly specifies to
+	// allow the insecure account unlocking, otherwise it is disabled.
+	if isHttp, ok := ctx.Value("http").(bool); ok && isHttp && !s.b.AccountManager().Config().InsecureUnlockAllowed {
```
Comment (karalabe) [inline internal/ethapi/api.go]: We need to tunnel in here **the** node, to check if http or ws is open at this point in time and deny based on that.

**Comment by holiman** [comment]
As it now works, I think it's fine. 

- If you call `unlock` via CLI, it fails the unlock if `ws` or `rpc` is enabled, 
- If you call `personal.unlock` via rpc, it fails it if `ws` or http-`rpc` is enanbled

It is still possible to shoot yourself in the foot if you really try hard, via 
```
> build/bin/geth --unlock 0 --password password.txt --nodiscover --maxpeers 0 console
[...]
> admin.startRPC()
INFO [03-29|11:32:57.269] HTTP endpoint opened                     url=http://localhost:8545         cors= vhosts=localhost
true
```

File: cmd/geth/main.go [hunk 11]
```diff
@@ -400,3 +393,21 @@ func startNode(ctx *cli.Context, stack *node.Node) {
 		}
 	}
 }
+
+// unlockAccounts unlocks any account specifically requested.
+func unlockAccounts(ctx *cli.Context, stack *node.Node) {
+	// If insecure account unlocking is not allowed if node's APIs are exposed to external.
+	// Print warning log to user and skip unlocking.
+	if !stack.Config().InsecureUnlockAllowed && stack.Config().ExtRPCEnabled() {
```
Comment (holiman) [inline cmd/geth/main.go]: Before we do this check, we might check if there are any unlocks to perform. Otherwise we'll always print the warning message

**Comment by rjl493456442** [comment]
@holiman updated

**Pull request merged**